### PR TITLE
[Merged by Bors] - Fix Bevy crashing if no audio device is found

### PR DESCRIPTION
**Pull request by forbjok**
Fixes https://github.com/bevyengine/bevy/issues/850

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

**Review by mockersf — APPROVED** [review]

File: crates/bevy_audio/src/audio_output.rs [hunk 1]
```diff
@@ -36,9 +44,11 @@ where
...
 {
     fn play_source(&self, audio_source: &P) {
-        let sink = Sink::try_new(&self.stream_handle).unwrap();
-        sink.append(audio_source.decoder());
-        sink.detach();
+        if let Some(stream_handle) = &self.stream_handle {
+            let sink = Sink::try_new(&stream_handle).unwrap();
+            sink.append(audio_source.decoder());
+            sink.detach();
+        }
```
Comment (NathanSWard) [inline crates/bevy_audio/src/audio_output.rs:51]: If you try to `play_source` when there is no audio device, I don't like the behavior of silentely succeeding. 
I would expect this to either
1) `error!`
2) `panic!`
3) have this return like `Result<(), ()>` (or a custom error type)

Comment (mockersf) [reply crates/bevy_audio/src/audio_output.rs:51]: From the player point of view, an `error` or `panic` are bad. "ok I know I don't have a audio device, don't crash and don't yell at me!". There's already a `warn` on the creation, re-logging it every time a sound play doesn't add much...

Comment (NathanSWard) [reply crates/bevy_audio/src/audio_output.rs:51]: That's fair. 
However, I have the counter argument that if you don't have an audio device, and therefore the `stream_handle` is `None`, if you are trying to play an `audio_source`, that's just wrong. Calling these functions if the type itself is not constructed successfully should not work. 

however, I also understand why the `panic!` and `warn!` would not be best for user experience.

Arguable, we should remove the `default` implementation, and instead do something like:
```rust
impl<P> AudioOutput<P> {
    pub fn new() -> Option<Self>;
}
```
However, I'm not actually sure if this is possible with how the value it initialized

Comment (MinerSebas) [reply crates/bevy_audio/src/audio_output.rs:51]: While `init_resource` requires `default` (actually `FromWorld`, but that has a blanket implementation for types that implement `default`), `insert_resource` doesn't have that bound.

An alternative is that `AudioOutput` is just never inserted as a Resource, and systems need to query for `Option<Res<AudioOutput>>` instead of `Res<AudioOutput>`

Comment (NathanSWard) [reply crates/bevy_audio/src/audio_output.rs:51]: > An alternative is that AudioOutput is just never inserted as a Resource, and systems need to query for Option<Res<AudioOutput>> instead of Res<AudioOutput>

Something like this would be my preferred solution :)

Comment (NathanSWard) [reply crates/bevy_audio/src/audio_output.rs:51]: I'm temporarily removing the `ready for cart` label until we decide on the best path forward from the above discussion.

Comment (forbjok) [reply crates/bevy_audio/src/audio_output.rs:51]: > if you don't have an audio device, and therefore the `stream_handle` is `None`, if you are trying to play an `audio_source`, that's incorrect

I disagree with that. The absence of an audio device isn't something I would really consider an error at all, it's just a possible circumstance on a user's computer. I would generally expect that if you run a game on a system that has no audio device, it should behave exactly the same as if you do have an audio device and just muted the volume. No sound, but the game runs perfectly fine otherwise.

Ideally, I would have preferred to be able to be able to disconnect and reconnect audio devices, or change the default output device at runtime and have it "just work" and play newly played sounds through the correct device, but that would likely require a bigger change. Currently, if you disconnect the only audio device while Bevy is running, the audio will just stop working and no more audio will be played even if you plug it back in, unless you restart the game.

Comment (NathanSWard) [reply crates/bevy_audio/src/audio_output.rs:51]: > The absence of an audio device isn't something I would really consider an error at all, it's just a possible circumstance on a user's computer

I don't disagree with this. :)
What I was saying is that if you don't have an audio device (which is ok) but then you try to play audio with a non existent audio device, that should probably not work. 
This is at least specific to the function level implementation.
If we want to support not panicking when a device (`AudioSource`) isn't available, that should be at the system level (as @MinerSebas recommended).

Comment (NathanSWard) [reply crates/bevy_audio/src/audio_output.rs:51]: Something like 
```rust
fn audio_system<T>(assets: Res<Assets<T>>, output: Option<Res<AudioOutput<T>>>, audio: Res<Audio<T>>) {
    if let Some(output) = output {
        output.try_play_queued(&*assets, &*audio);
    }
}
```

(or something to this extent)

Comment (NathanSWard) [reply crates/bevy_audio/src/audio_output.rs:51]: However, if people really feel strongly about the current implementation, I'm more than happy to mark this as resolved.

Comment (forbjok) [reply crates/bevy_audio/src/audio_output.rs:51]: > An alternative is that `AudioOutput` is just never inserted as a Resource, and systems need to query for `Option<Res<AudioOutput>>` instead of `Res<AudioOutput>`

The issue with this is that it would require users of Bevy to have an `if let Some(audio_output) = audio output` in every place where audio is being played. This seems very inelegant and clunky. I think it's much better to have AudioOutput itself always exist, and have it handle the existence (or non-existence) of devices internally so that users don't have to worry about it.

I experimented a bit with changing it so that the streams are stored in a RefCell (initialized to None by default) and then initialized on demand, and while this basically works and solves the problem of Bevy crashing, I wasn't able to find a way to detect when the stored OutputStreamHandle has broken due to the device having been disconnected, so being able to recover from that might require changes in `rodio`.

Comment (NathanSWard) [reply crates/bevy_audio/src/audio_output.rs:51]: >  I think it's much better to have AudioOutput itself always exist, and have it handle the existence

Sure, and when you phrase it like that I don't mind it as much :)
Feel free to marked the thread as resolved and thanks for the discussion 👍

Comment (Pauan) [reply crates/bevy_audio/src/audio_output.rs:51]: > The issue with this is that it would require users of Bevy to have an `if let Some(audio_output) = audio` output in every place where audio is being played. This seems very inelegant and clunky. 

Using `if let` is a *very* common pattern in Rust. Unlike most other languages, Rust has a *very* strong priority on correctness, it really pushes you to manually handle failure conditions rather than silently ignoring them.

Rust was never designed as a loose scripting language where you can just do whatever and have it "just work", and trying to make it into a loose scripting language will have a lot of long term negative effects.

I can also see plenty of situations where it is *mandatory* to have audio, or where the programmer wants to handle audio failure in a specific way. Using `Option` to represent a missing resource or component is correct.

> The absence of an audio device isn't something I would really consider an error at all, it's just a possible circumstance on a user's computer

That is correct, it is not an error, but it *is* a *failure*. Rust makes a clear distinction between failures and errors. If a key is missing in a dictionary, that is a failure (which is why [`HashMap::get`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.get) returns an `Option`). Rust has very good systems for handling failures (`if let`, `match`, `?`, `try`, and combinators).

Other languages also have a distinction between [errors and failures](https://www.joyent.com/node-js/production/design/errors), but this distinction is implicit (not explicit), and many programmers are unaware of the distinction. In Rust the distinction is explicit (`Option` vs `Result` vs panic).

Comment (forbjok) [reply crates/bevy_audio/src/audio_output.rs:51]: > I can also see plenty of situations where it is _mandatory_ to have audio, or where the programmer wants to handle audio failure in a specific way.

For these most likely extremely rare cases, I think it would make more sense to have a way to explicitly check whether an audio device is present.

> That is correct, it is not an error, but it _is_ a _failure_. Rust makes a clear distinction between failures and errors.

I would argue that from the perspective of a user of the engine (game developer), as well as end-users, trying to play a sound when there is no audio device connected is not even a failure. For all intents and purposes, it's the same as if you play audio through a sound device that is muted or doesn't have a speaker plugged in. You obviously don't get any sound, but it is not in any way wrong of the game to try to play sound.

Comment (Pauan) [reply crates/bevy_audio/src/audio_output.rs:51]: > For these most likely extremely rare cases, I think it would make more sense to have a way to explicitly check whether an audio device is present.

Yes, that is exactly what `Option` is. And the case is not "extreme", it is quite common for games to inform the user about missing hardware.

> This argument makes sense in the context of a low-level audio library, but not so much in the context of a game engine.

Not at all, the distinction is pervasive at all levels, it has nothing to do with "low level" vs "high level". Every language (including high level scripting languages like JavaScript) have this concept. The link in my previous comment explains this in more detail. It is simply a fundamental programming principle.

> I would argue that from the perspective of a user of the engine (game developer), as well as end-users, trying to play a sound when there is no audio device connected is not even a failure.

In some cases that may be true, but in some cases it is not. That's why it should be `Option`, so the programmer can *choose* what behavior they want:

* They can ignore the failure.
* They can propagate the failure to somewhere else.
* They can crash.
* They can report the failure in some way.
* Some combination of the above.

Using `Option` gives complete control to the programmer.

> For all intents and purposes, it's the same as if you play audio through a sound device that is muted or doesn't have a speaker plugged in.

That's not true. If the audio is muted that is an explicit choice by the user. If the audio is failing, that is generally *not* the user's choice, instead it is caused by some sort of bug, hardware failure, or misconfiguration.

The game might wish to display a popup box to the user saying "Could not find audio device, make sure your audio is configured properly". If the game does *not* inform the user of the problem, then the user will wonder, "why is my audio not working?"

Comment (NathanSWard) [reply crates/bevy_audio/src/audio_output.rs:51]: I think something also to consider for right now is that the current standing of `bevy_audio` is not the best. 
It'll mostly likely get replaced in the future via `bevy_kira_audio` so I think right now, it should be our goal to make the built-in functionality as simple and friendly as possible, as the API will likely get scrapped and reworked completely.

Comment (Pauan) [reply crates/bevy_audio/src/audio_output.rs:51]: @NathanSWard That's a completely fair position, and if it eventually gets done Correctly™ then that's fine, and I agree that a proper solution should be more general (for all resources/systems), so it's okay to punt it to a later PR.

But there's a big difference between, "this is hacky and we know it's hacky and we'll do it correctly later" and "this is a good design which we should keep". I'm advocating for the first, not the second.

Comment (mockersf) [reply crates/bevy_audio/src/audio_output.rs:51]: > I can also see plenty of situations where it is _mandatory_ to have audio, or where the programmer wants to handle audio failure in a specific way. Using `Option` to represent a missing resource or component is correct.

Just wondering, and pretty much out of topic, sorry about that... I pretty much always play with no sound because of reasons. Should the volume being off be detected as a failure then?

Comment (Pauan) [reply crates/bevy_audio/src/audio_output.rs:51]: > Should the volume being off be detected as a failure then?

There's two basic designs:

1. Treat it as not a failure, and have a separate `is_muted` method.
2. Return a `Result` which has an `enum` that describes the various failures that can happen, with `Muted` being one of the reasons.

Either design is fine, they just make different tradeoffs. Personally I would go with the first one.

Comment (cart) [reply crates/bevy_audio/src/audio_output.rs:51]: I agree with the idea that users shouldn't need to care about this (most of the time). I personally don't think the spirit of rust is "whenever you have a chance to require explicit error handling ... do it". I think in rust, just like every other language, you should make a pragmatic call based on the context.

Ex: we _could_ make everyone explicitly handle the case where they try to spawn a sprite, but a render device isn't present by making `spawn().insert(Sprite)` return a RenderDeviceNotAvailable error. But I would argue that doing that would "get in the way" more often than not. A missing render device is a corner case. If every api involving "rendering things" required checking for render device existence, users would have a "very bad time". And when its not there, its generally because the user intended it to be that way (ex; they are running in a headless environment). This is why we have a "headless render backend". We handle this "internally" so when someone tries using sprites, they don't need to think about render devices / account for them being missing.

I would hazard a guess that in the 99.X% case, when users play an audio clip, they don't want to care about audio device existence via `if let` / `unwrap`. Therefore we should have an api that optimizes for that use case. We _should_ still expose "device existence information" to users that care about it. But we constantly make choices to hide information / assume a certain "happy path" state on behalf of our users in the interest of ergonomics and a "pleasant programming experience". This feels like a case where we should do that.

I'm happy to continue this discussion, but I think the current impl is a solid short term solution to a nasty problem. And it has the benefit of not degrading UX for a common use case (and requiring `if let Some(audio_device)` for all audio operation definitely degrades UX for a common use case).
